### PR TITLE
fixed memory bugs, and all arrays of arbitrary length are supported now as well as some minor bug fixes

### PR DESCRIPTION
**Pull request by NightMare-Vortex**
Sorry for my shameless dependency plug, btw

### CODE REVIEW

File: src/impls/simdjson.rs [hunk 1]
```diff
@@ -89,16 +100,30 @@ impl<'input, 'tape> BorrowedDeser<'input, 'tape> {
...
+            res.push(self.parse().unwrap())
         }
+
+        // Sorry but the bellow code can cause memory UB
+        // if parse panics, or if unwrap panics
+        // now lets say some uninit value within the array was
+        // was Value::String(Cow::Owned(some_string)))
+        // now res.drop() -> for i in 0..res.len (*res.ptr).drop() -> calls uninit.drop() -> dealloc(unowned memory)
+        // and there you go freeing memory you dont own
+        // not only that but parse_array can fail with unexpected Eof
```
Comment (Licenser) [inline src/impls/simdjson.rs:114]: This function translates the tape, which is validated and parsed, to an array we know that `len` elements exist and are valid.

I have to look into why parse() returns are result and not is a always valid function

**Review by Licenser — APPROVED** [review]
I quite like the changes, nice QOL improvements 👍 .

In general, I would prefer the Deserializer/Serializer to live in the `heap-array` crate since there are more data structure crates then there are encoding crates but since the work is done here already I don't mind merging it. :)

File: src/impls/collections.rs [hunk 2]
```diff
@@ -206,37 +211,44 @@ where
...
+                }
             }
-            writer.write_all(b"}")
-        } else {
-            writer.write_all(b"{}")
         }
-    }
+    };
+}
+macro_rules! de_map_like {
```
[the remark this answers is not in the document]
Comment (NightMare-Vortex) [reply src/impls/collections.rs:242]: uh yeah I was trying to do some code clean up with the deserialization of maps, but It really didnt make sense to have the macro take in the deserializer and I forgot to remove it from the final code

**Comment by NightMare-Vortex** [comment]
Thanks for the code review, I just don't know how we proceed now

**Comment by NightMare-Vortex** [comment]
> I quite like the changes, nice QOL improvements 👍 .
> 
> In general, I would prefer the Deserializer/Serializer to live in the `heap-array` crate since there are more data structure crates then there are encoding crates but since the work is done here already I don't mind merging it. :)

It's fine I moved the Deserializer/Serializer to live in the `heap-array`

**Comment by Licenser** [comment]
Heya @NightMare-Vortex I updated it to resolve conflicts for the 0.11 release, it'll still need a `cargo fmt` run before it can be merged

**Comment by Licenser** [comment]
Heya @NightMare-Vortex  are you still interested in finishing this PR?

**Comment by NightMare-Vortex** [comment]
> Heya @NightMare-Vortex are you still interested in finishing this PR?

oh yes I am, sorry I didn't see anything on my email 😅

**Comment by Licenser** [comment]
No worries, I know how easy it is to get flooded with mails :D that's why I figured I'll re-ping every now and then

File: src/lib.rs [hunk 3]
```diff
@@ -70,15 +73,31 @@ impl<T: AsRef<str>> SerializeAsKey for T {
...
 
     #[inline]
     fn from_slice(json: &'input mut [u8]) -> simd_json::Result<Self>
     where
-        Self: std::marker::Sized + 'input,
+        Self: Sized + 'input,
     {
         let tape = simd_json::to_tape(json)?;
-        let mut itr = tape.0.into_iter().peekable();
+        let mut itr = tape.into_iter().peekable();
```
Comment (Licenser) [inline src/lib.rs:84]: there seems to be merge conflicts here

**Review by Licenser — APPROVED** [review]
Looks over all good 👍 some merge conflicts snuck in and fail the CI otherwise 🚀

**Review by Licenser — APPROVED** [review]
<3

**Comment by Licenser** [comment]
As there were a bunch of conflicts I cherry picked the changes and merged them in #72

**Comment by NightMare-Vortex** [comment]
Alright thanks

**Comment by NightMare-Vortex** [comment]
although I really do think there should be a discussion about the safety of the parsing code, why is parse fallible, and why we parse the array, without having the values of the array in the first place

**Pull request closed without merging**